lib/fizzy/parser_expr.cpp

```diff
@@ -63,17 +71,27 @@ parser_result<Code> parse_expr(const uint8_t* pos, const uint8_t* end, bool have
     // For a block/if/else instruction the value is the block/if/else's immediate offset.
     std::stack<ControlFrame> control_stack;
 
-    control_stack.push({Instr::block, 0});  // The function's implicit block.
+    control_stack.push({Instr::block});  // The function's implicit block.
+
+    const auto metrics_table = get_instruction_metrics_table();
 
     bool continue_parsing = true;
     while (continue_parsing)
     {
         if (pos == end)
             throw parser_error{"Unexpected EOF"};
 
-        const auto& frame = control_stack.top();
+        auto& frame = control_stack.top();
+
+        const auto opcode = *pos++;
+        const auto& metrics = metrics_table[opcode];
+
+        if (frame.stack_height < metrics.stack_height_required && !frame.unreachable)
+            throw validation_error{"stack underflow"};
+
+        frame.stack_height += metrics.stack_height_change;
```

Comment:
Should this be increased when `frame.unreachable`?

Reply:
I think it's simpler to always increase without condition, otherwise we'd need to also add condition when increasing it in `loop`, `block` and `if_`